### PR TITLE
[MNG-8018] Fix MSYS/Git-Bash path mis-detection on Windows

### LINKED ISSUE #10834: [MNG-8018] Maven creates local repo at wrong path
**[Jörg Hohwiller](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hohwille)** opened **[MNG-8018](https://issues.apache.org/jira/browse/MNG-8018?redirect=false)** and commented

I am hunting bugs with maven for years now that is creating a new local repository in an odd location (something like `D:\d\...` but before also `$PWD/${M2_REPO`} - see https://github.com/devonfw/ide/issues/706). Now I was finally able to become able to effectively reproduce it:

I created an empty `d` folder on my `D:` drive but removed all permissions on it even for myself to force errors in case maven will ever try to populate this directory again.
This now happened when maven is executed from `maven-exec-plugin` and the nested maven tries to resolve the local repository in that context it seems to consider the path to the local repository as relative even though it is absolute.
Here are some more detailed circumstances when this error is happening:
* I am on Windows
* I am working in git-bash when I invoke maven (`mvn install`).
* To relocate the local repository, I used this "hack": `export MAVEN_OPTS= -Duser.home=/d/projects/mmm/conf`. This works to take the settings from `/d/projects/mmm/conf/.m2/settings.xml` what in my case configures my repo to be at `/c/Users/hohwille/conf/.m2/repository` (git-bash syntax) or `C:\Users\hohwille\\.m2\\repository` (Windows syntax) at least for the top-level `mvn` process.
* Now when I did run `mvn install` the project was executing `mvn install` in a test folder via `exec-maven-plugin`. See https://github.com/m-m-m/code/blob/238899e2b2525081b39f59932e093d5c331a2864/java/impl/pom.xml#L64
  This results in the following error:

```
[INFO] --- exec:3.1.0:exec (Install test project) @ mmm-code-java-impl ---
[ERROR] Could not create local repository at D:\d\projects\mmm\conf\.m2\repository -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/LocalRepositoryNotAccessibleException
[ERROR] Command execution failed.
org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1)
    at org.apache.commons.exec.DefaultExecutor.executeInternal (DefaultExecutor.java:404)
    at org.apache.commons.exec.DefaultExecutor.execute (DefaultExecutor.java:166)
    at org.codehaus.mojo.exec.ExecMojo.executeCommandLine (ExecMojo.java:1000)
    at org.codehaus.mojo.exec.ExecMojo.executeCommandLine (ExecMojo.java:947)
    at org.codehaus.mojo.exec.ExecMojo.execute (ExecMojo.java:471)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
```

I was already hunting down the source-code of maven-exec plugin if it contains any code related to this problem. But it seems not to explicitly handle or tweak any related environment variables or do anything different where `mvn` is executed or any other program.
So my assumption is that maven core itself has some logic to resolve the local maven repo and decide if something is a relative or an absolute path. As this is working for paths in git-bash syntax for the top-level `mvn` process, I would guess that there is some magic detection of git-bash/MSYS that however, stops working if `mvn` is then again executed but not directly from git-bash but from a Java-process (maven itself) within.
BTW: With the path treated wrongly as relative it can not find the `settings.xml` and hence computes the path to the local repository in the "default location" what is inside my tweaked home directory causing the difference between `C` and `D` drive.

I know this is kind of an edge-case but as I was searching for this for a long time, I decided to report this. As I could not see it is a bug of `maven-exec-plugin` itself, I finally filed it here. I would be very pleased, if you have any further ideas on this.
In case I should try anything special please let me know and I will do and provide results.
You could however easily reproduce this by creating a simple project with again a very simple maven project in `src/test/resources/maven.project`
such pom section:

```
...
  <build>
    <plugins>
      <plugin>
        <groupId>org.codehaus.mojo</groupId>
        <artifactId>exec-maven-plugin</artifactId>
        <version>3.1.0</version>
        <executions>
          <execution>
            <id>Install test project</id>
            <phase>test-compile</phase>
            <goals>
              <goal>exec</goal>
            </goals>
          </execution>
        </executions>
        <configuration>
          <executable>mvn</executable>
          <workingDirectory>target/test-classes/maven.project</workingDirectory>
          <arguments>
            <argument>install</argument>
          </arguments>
        </configuration>
      </plugin>
    </plugins>
  </build>
  ...
```

Then in `git-bash` run `export MAVEN_OPTS="-Duser.home=$HOME" && mvn install`. The secondary `mvn` process should then create a repo in something like `C:\c\Users\«yourlogin»\.m2\repository`.

---

**Affects:** 3.9.6

**Attachments:**
- [mng-8018.tgz](https://issues.apache.org/jira/secure/attachment/13066667/mng-8018.tgz) (_1.18 kB_)

### PR DESCRIPTION
**Pull request by arturobernalg**
Fixes https://github.com/apache/maven/issues/10834

Adds JUnit 5 Windows-only test for path normalisation.

Following this checklist to help us incorporate your
contribution quickly and easily:

- [X ] Your pull request should address just one issue, without pulling in other changes.
- [ X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
  Note that commits might be squashed by a maintainer on merge.
- [ X] Write unit tests that match behavioral changes, where the tests fail if the changes to the runtime are not applied.
  This may not always be possible but is a best-practice.
- [ X] Run `mvn verify` to make sure basic checks pass.
  A more thorough check will be performed on your pull request automatically.
- [ X] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

- [ X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
- [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

File: compat/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java [hunk 1]
```diff
@@ -174,6 +174,25 @@ public class MavenCli {
 
     private static final Pattern NEXT_LINE = Pattern.compile("\r?\n");
 
+    /** Matches “/d/whatever” or “/C/…”.  */
+    private static final Pattern MSYS_PATH = Pattern.compile("^/([a-zA-Z])/(.*)$");
+
+    /**
+     * Turns “/d/projects/foo” into “D:\\projects\\foo”.
+     * Returns the original string if it isn't an MSYS-style path.
```
Comment (elharo) [inline compat/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java]: What's MSYS?

File: compat/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java [hunk 2]
```diff
@@ -174,6 +174,25 @@ public class MavenCli {
...
+     */
+    static String msysToWindowsPath(final String p) {
```
Comment (elharo) [inline compat/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java]: p --> path

File: compat/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java [hunk 3]
```diff
@@ -174,6 +174,25 @@ public class MavenCli {
...
+        if (p == null) {
+            return null;
+        }
+        final Matcher m = MSYS_PATH.matcher(p);
```
Comment (elharo) [inline compat/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java]: m --> matcher

File: compat/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java [hunk 4]
```diff
@@ -363,6 +382,22 @@ void initialize(CliRequest cliRequest) throws ExitException {
         //
         String mavenHome = System.getProperty(Constants.MAVEN_HOME);
 
+        if (org.codehaus.plexus.util.Os.isFamily("windows")) {
+            System.setProperty("user.home", msysToWindowsPath(System.getProperty("user.home")));
+
+            for (final String k : new String[] {
```
Comment (elharo) [inline compat/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java]: k?

File: compat/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java [hunk 5]
```diff
@@ -363,6 +382,22 @@ void initialize(CliRequest cliRequest) throws ExitException {
...
+                Constants.MAVEN_REPO_LOCAL, // -Dmaven.repo.local
+                "user.settings", // -s / --settings
+                "alternateSettings",
+                "user.toolchains"
+            }) {
+                final String v = System.getProperty(k);
```
Comment (elharo) [inline compat/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java]: avoid single letter variable names

**Review by elharo — CHANGES_REQUESTED** [review]

File: compat/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java (hunk 1 above)
Comment (arturobernalg) [reply compat/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java]: Minimal System, is a software distribution and build platform for Windows

Comment (elharo) [reply compat/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java]: MSYS2?

Might need to spell this out. Is MSYS really the style of string here, or is this just a Windows path?

Comment (arturobernalg) [reply compat/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java]: POSIX-style Windows paths produced by MSYS2/Git Bash (and Cygwin), e.g. /c/Users/... or /cygdrive/c/.... The method converts those to native C:\... and returns the input if it’s already a native Windows path.

**Review by gnodet — COMMENTED** [review]
## Review: [MNG-8018] Fix MSYS/Git-Bash path mis-detection on Windows

The fix addresses a real bug, but the `LookupInvoker` implementation has several issues that need attention before merging.

### High

1. **`System.out.println` in production code** (`LookupInvoker.normalizeMsysPath`) — A debug statement `System.out.println("Normalized MSYS path ...")` is left in the method, right after the proper `context.logger.debug(...)` call. This will pollute console output for every Windows user with a `/`-prefixed path. Remove it.

2. **Over-broad path detection** (`LookupInvoker.normalizeMsysPath`) — The method treats *any* path starting with `/` as an MSYS drive-letter path and blindly extracts `path.substring(1, 2)` as the drive letter. A path like `/tmp/foo` would be converted to `T:\mp\foo`. The compat `MavenCli.msysToWindowsPath` correctly uses a regex `^/([a-zA-Z])/(.*)$` that requires a single letter followed by `/`. Port that approach here.

### Medium

3. **Broad `catch (Exception)`** — The catch-all masks `StringIndexOutOfBoundsException` for edge cases like a path that is just `"/"`. This is exception-based flow control. The regex approach from MavenCli avoids this entirely.

4. **Two divergent implementations** — `MavenCli.msysToWindowsPath` (regex-based, robust) and `LookupInvoker.normalizeMsysPath` (substring-based, fragile) implement the same logic differently. Consider extracting to a shared utility.

5. **Incomplete property normalization** — `LookupInvoker.invoke` only normalizes `user.home`. The compat `MavenCli.initialize` also normalizes `maven.repo.local`, `user.settings`, `alternateSettings`, and `user.toolchains`. The impl version should cover the same properties for parity.

### Low

6. **Javadoc claims Cygwin support** (`MavenCli.msysToWindowsPath`) — The Javadoc mentions `/cygdrive/c/...` paths, but the regex `^/([a-zA-Z])/(.*)$` does not match that pattern. Either add Cygwin support or correct the Javadoc.

7. **Integration tests are Windows-only** — `@EnabledOnOs(OS.WINDOWS)` means these tests never run on Linux CI. The pure conversion logic could have a platform-independent unit test (like `MavenCliTest.msysPathsAreConverted` does).

8. **`System.setProperty("user.home")` is a global JVM mutation** — Both implementations modify this. Mitigated by the existing save/restore pattern in `invoke`, but worth a comment.

---

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

File: impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/LookupInvoker.java [hunk 6]
```diff
@@ -918,6 +931,29 @@ protected int calculateDegreeOfConcurrency(String threadConfiguration) {
...
+    // Helper method to normalize MSYS-style paths on Windows
+    private String normalizeMsysPath(String path, C context) {
+        if (path != null && isWindows() && path.startsWith("/")) {
+            try {
+                // Convert /d/path to D:\path
+                final String drive = path.substring(1, 2).toUpperCase(Locale.ENGLISH);
+                final String rest = path.substring(2).replace("/", "\\");
+                final String normalized = drive + ":" + rest;
+                context.logger.debug("Normalized MSYS path '" + path + "' to '" + normalized + "'");
+                System.out.println("Normalized MSYS path '" + path + "' to '" + normalized + "'");
```
Comment (gnodet) [inline impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/LookupInvoker.java:943]: Debug leftover — `System.out.println` must be removed. The line above already uses `context.logger.debug(...)` which is the correct approach.

File: impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/LookupInvoker.java [hunk 7]
```diff
@@ -918,6 +931,29 @@ protected int calculateDegreeOfConcurrency(String threadConfiguration) {
                     + "'. Supported are int and float values ending with C.");
         }
     }
+    // Helper method to normalize MSYS-style paths on Windows
+    private String normalizeMsysPath(String path, C context) {
```
Comment (gnodet) [inline impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/LookupInvoker.java:935]: Bug: This converts ANY path starting with `/` to a Windows drive path, not just MSYS drive-letter paths. For example, `/tmp/repo` would produce `T:mp\repo`.

The method uses `path.substring(1, 2)` without verifying that position 2 is `/`. `MavenCli.java` correctly uses the regex `^/([a-zA-Z])/(.*)$` which requires exactly a single letter between two slashes.

Consider reusing the same regex approach:
```java
private static final Pattern MSYS_PATH = Pattern.compile("^/([a-zA-Z])/(.*)$");

private String normalizeMsysPath(String path, C context) {
    if (path != null && isWindows()) {
        Matcher matcher = MSYS_PATH.matcher(path);
        if (matcher.matches()) {
            String normalized = Character.toUpperCase(matcher.group(1).charAt(0))
                    + ":\\" + matcher.group(2).replace('/', '\\');
            context.logger.debug("Normalized MSYS path '" + path + "' to '" + normalized + "'");
            return normalized;
        }
    }
    return path;
}
```

**Review by gnodet — CHANGES_REQUESTED** [review]
The PR addresses a real MSYS/Git Bash path mis-detection bug (MNG-8018), and the `MavenCli.java` implementation is well done with proper regex validation. However, the `LookupInvoker.java` implementation has critical correctness issues. All findings independently verified (7/7 confirmed).

**Critical:**
1. **Debug leftover (high)** — `System.out.println("Normalized MSYS path...")` at LookupInvoker.java:943 must be removed. Maven uses its own logging infrastructure.
2. **Over-broad path conversion (high)** — `normalizeMsysPath` converts ANY path starting with `/` to a Windows drive path, not just MSYS drive-letter paths. For example, `/tmp/repo` produces `T:mp\repo`. The method uses `path.substring(1, 2)` without verifying the character at position 2 is `/`. `MavenCli.java` correctly uses the regex `^/([a-zA-Z])/(.*)$`.

**Medium:**
3. **Code duplication** — Two divergent implementations: `MavenCli.msysToWindowsPath` (regex-based, correct) vs `LookupInvoker.normalizeMsysPath` (substring-based, buggy). Consider a shared utility.
4. **Misleading Javadoc** — `msysToWindowsPath` Javadoc claims Cygwin `/cygdrive/c/...` support but the regex doesn't match that pattern.
5. **Missing null check** — `isWindows()` calls `System.getProperty("os.name").toLowerCase()` without null guard.

**Low:**
6. Missing blank line between methods.
7. Test covers happy path only — no edge cases (null, non-MSYS paths, empty string).

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of gnodet_

File: impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/LookupInvoker.java (hunk 6 above)
Comment (gnodet) [inline impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/LookupInvoker.java:943]: Debug leftover — `System.out.println` must be removed. The line above already uses `context.logger.debug(...)` which is the correct approach.

File: impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/LookupInvoker.java (hunk 7 above)
Comment (gnodet) [inline impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/LookupInvoker.java:935]: Bug: This converts ANY path starting with `/` to a Windows drive path, not just MSYS drive-letter paths. For example, `/tmp/repo` would produce `T:mp\repo`.

The method uses `path.substring(1, 2)` without verifying that position 2 is `/`. `MavenCli.java` correctly uses the regex `^/([a-zA-Z])/(.*)$` which requires exactly a single letter between two slashes. Consider reusing the same regex approach.

**Review by gnodet — CHANGES_REQUESTED** [review]
The PR addresses a real MSYS/Git Bash path mis-detection bug (MNG-8018), and the `MavenCli.java` implementation is well done with proper regex validation. However, the `LookupInvoker.java` implementation has critical correctness issues. All findings independently verified (7/7 confirmed).

**Critical:**
1. **Debug leftover (high)** — `System.out.println("Normalized MSYS path...")` at LookupInvoker.java:943 must be removed. Maven uses its own logging infrastructure.
2. **Over-broad path conversion (high)** — `normalizeMsysPath` converts ANY path starting with `/` to a Windows drive path, not just MSYS drive-letter paths. For example, `/tmp/repo` produces `T:mp\repo`. The method uses `path.substring(1, 2)` without verifying the character at position 2 is `/`. `MavenCli.java` correctly uses the regex `^/([a-zA-Z])/(.*)$`.

**Medium:**
3. **Code duplication** — Two divergent implementations: `MavenCli.msysToWindowsPath` (regex-based, correct) vs `LookupInvoker.normalizeMsysPath` (substring-based, buggy). Consider a shared utility.
4. **Misleading Javadoc** — `msysToWindowsPath` Javadoc claims Cygwin `/cygdrive/c/...` support but the regex doesn't match that pattern.
5. **Missing null check** — `isWindows()` calls `System.getProperty("os.name").toLowerCase()` without null guard.

**Low:**
6. Missing blank line between methods.
7. Test covers happy path only — no edge cases (null, non-MSYS paths, empty string).

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of gnodet_

**Comment by slachiewicz** [comment]
The last CI run on this branch was 2025-08-11, and GitHub has deleted the logs since, so there's no way to tell whether the change still builds against master.

A rebase triggers a fresh run and makes the state readable again. If you'd rather not carry it forward, closing is equally useful — the PR isn't being ignored, it's unreadable.

The branch also conflicts with master now, so that needs resolving before CI can say anything. Say the word if you'd like a hand with it.

*This comment was created with AI assistance.*
